ACE codegen refactor

adr1anh opened the pull request:
The goal of this PR is to improve the code generation of an ACE circuit given an AIR, and make it easier to generate code for a full MASM verifier, by making it easier to obtain offsets of chunks of inputs, and improve hashing efficiency by aligning input sections.

- Add `InputLayout` struct.
  - Adds the ability to pad different input sections to be (double-)word aligned. This requires the proof to be padded as well, which in turn requires Winterfell support.
  - The layout is currently different from the one established previously, but this can be changed by modifying the layout constructor.
  - Currently, we align the OOD evaluations of each trace to simplify the Horner evaluation for the DEEP quotient. 
  - Supports the idea of evaluating the quotient at the next point, to unify the linear combination of the OOD points across main and next rows. This can be made optional while waiting for Winterfell support.
  - Auxiliary STARK inputs are given a fixed structure using an enum, simplifying their layout and making it easier to lay out in memory

- Copy circuit implementation from the VM, using own types, simplifying padding
  - The circuit is constructed by considering inputs, constants, and operations separately, as this simplifies padding when we need to include dummy inputs, dummy constants or instructions
  - We create the circuit from the AIR representation by making use of the assumption that the nodes in the AIR graph are already sorted, requiring only a linear pass over the graph nodes

- Update circuit to discard the last point of the trace

- Move testing airs to separate folder, avoiding the need to export structs used only for testing.
  - Test circuit construction by generating random inputs, and provide native quotient evaluation. By using valid inputs previously, constraints would evaluate to zero, therefore, the numerator would be zero. This ensures the native and circuit computations are equivalent.
  - Removed previous manual derivation of constraints, thereby simplifying testing code.

- Move all dot generation to separate file
  - implement `Display` for nodes to be more descriptive
  - Improve dot description of different node types

- Move encoding to separate file
  - Hashing is still missing but will be added shortly 
  - Avoids embedding the circuit sizes as header, as these can be embedded in the MASM generated code


- Add operation optimizations for circuit builder
  - constant evaluation
  - create circuit by recursive depth-first search, ensuring proper ordering
  - avoid unnecessary last alpha computation

- Improve exporting of types

Al-Kindi-0 commented on codegen/ace/src/encoded.rs:
````diff
@@ -0,0 +1,280 @@
+    /// ```
+    pub fn to_ace(&self) -> EncodedCircuit {
+        const MAX_NODE_ID: u64 = (1 << 30) - 1;
+        // assert!(self.is_padded(), "circuit not padded");
````
should we remove?

Al-Kindi-0 commented on codegen/ace/src/encoded.rs:
```diff
@@ -0,0 +1,280 @@
+                .iter()
+                .flat_map(|c| QuadFelt::from(*c).to_base_elements()),
+        );
+        // Since constants are treated as extension field elements, we add an extra 0
+        // to ensure the section in the circuit is word-aligned
```
nit: "... add extra 0-s" or something similar

Al-Kindi-0 commented on codegen/ace/src/encoded.rs:
```diff
@@ -0,0 +1,280 @@
+                Node::Constant(idx) => constants_start.checked_sub(idx),
+                Node::Operation(idx) => ops_start.checked_sub(idx),
+            }
+            .unwrap() as u64
```
we should probably use "expect"

Al-Kindi-0 commented on codegen/ace/src/encoded.rs:
```diff
@@ -0,0 +1,280 @@
+            let id_1 = node_id(operation.node_r);
+            let op_tag = operation.op as u64;
+
+            const OP_OFFSET: u64 = 1 << 60;
+            const ID_1_OFFSET: u64 = 1 << 30;
```
Do we need these elsewhere? If so, it might be worth it to define them outside

adr1anh replied:
Not entirely sure what to do with them. We could make them public and import them from the VM, but this could be done when we import the AirScript from the VM with the generated verifier.

Al-Kindi-0 reviewed (approved):
Looks great!
Left some minor comments.

Al-Kindi-0 commented:
Could we also rebase on the most recent commit?

adr1anh commented on codegen/ace/src/builder.rs:
```diff
@@ -0,0 +1,482 @@
+/// - no dangling references (e.g., a node pointing to a child that does not exist)
+/// - no cycles (the circuit is a DAG)
+/// - no disconnected nodes, except possibly for inputs
+/// - no duplicated nodes
```
Done

adr1anh commented on codegen/ace/src/builder.rs:
```diff
@@ -0,0 +1,482 @@
+        self.push_op(ArithmeticOp::Sub, node_l, node_r)
+    }
+
+    /// Returns the sum of [`Node`]s returned by the `els` `Iteration`
```
Done

adr1anh commented on codegen/ace/src/encoded.rs:
```diff
@@ -0,0 +1,280 @@
+impl Circuit {
+    /// Serializes a [`Circuit`] to a list of field elements in the format expected by the
+    /// ACE chiplet.
+    /// Note that some information is lost, so the original circuit cannot be
+    /// recovered from its felt serialization.
```
Removing the line because it's not entirely accurate. Yes the conversion is "lossy", but the original circuit descript could be recovered given its `AceInputs`.

adr1anh commented on codegen/ace/src/encoded.rs:
```diff
@@ -0,0 +1,280 @@
+    }
+
+    /// Number of input nodes
+    pub fn num_inputs(&self) -> usize {
+        self.num_vars - self.num_constants()
```
Explained in more detail in the docs for `EncodedCircuit`

adr1anh commented on codegen/ace/src/encoded.rs:
```diff
@@ -0,0 +1,280 @@
+    ///
+    /// The output indexes of inputs, constants and gates are not serialized as they
+    /// are simply successive values starting from `n_inputs` to
+    /// `n_inputs + n_constants + n_gates - 1`.
```
Made the comment clearer about how the indices are mapped.

adr1anh commented on codegen/ace/src/inputs.rs:
```diff
@@ -0,0 +1,166 @@
+}
+
+/// Set of all variables related to the STARK IOP. These are not used by the constraints but are 
+/// necessary to evaluate the composition polynomial from the evaluations of the constraints.
```
I've addressed it, but I simplified the way the AceInputs are derived from the values provided by the user.

adr1anh commented on codegen/ace/src/inputs.rs:
```diff
@@ -0,0 +1,166 @@
+    /// The [`Air`] is required to compute `zᵐᵃˣ`, the power of `z` at which we evaluate the longest
+    /// periodic column, and from which we derive the evaluation points of all other columns.
+    pub fn new(air: &Air, log_trace_len: usize, alpha: QuadFelt, z: QuadFelt) -> Self {
+        let gen_max = Felt::TWO_ADIC_ROOT_OF_UNITY;
+        let gen_log_pow = Felt::TWO_ADICITY - log_trace_len as u32;
+        let gen: Felt = gen_max.exp_vartime(1 << gen_log_pow).into();
```
I didn't find it in winterfell.

Al-Kindi-0 replied:
https://github.com/facebook/winterfell/blob/6b665a6bde2bb064b2ee33e99b2a285174d1ab97/math/src/field/traits.rs#L258

adr1anh commented on codegen/ace/src/layout.rs:
```diff
@@ -0,0 +1,213 @@
+            .collect();
+
+        let random_values = next_region(&mut inputs_offset, air.num_random_values as usize);
+
+        // TODO(@adr1anh): Use the following to derive the degree generically.
+        // let degree = [0, 1]
+        //     .into_iter()
+        //     .flat_map(|segment| air.integrity_constraint_degrees(segment.into()))
+        //     .map(|deg| deg.base())
+        //     .max()
+        //     .unwrap();
+        //
+        // let num_quotient_elements = (degree - 1)
+        //     .next_power_of_two()
+        //     .next_multiple_of(HASH_ALIGNMENT);
+        // let quotient_degree = Self::quotient_parts(air);
```
Yes good idea, I'll create an issue but leave the TODOs for now

Al-Kindi-0 replied:
Same comment here

adr1anh replied:
#391

adr1anh commented on codegen/ace/src/lib.rs:
```diff
@@ -1,21 +1,148 @@
 use miden_core::{Felt, QuadExtension};
 
-pub struct CodeGenerator {}
+type QuadFelt = QuadExtension<Felt>;
```
I think we use `QuadFelt` everywhere now, as this was what the VM uses I think.

adr1anh commented on codegen/ace/src/tests/inputs.rs:
```diff
@@ -0,0 +1,201 @@
+                .integrity_constraints(segment)
+                .iter()
+                .map(|c| {
+                    // TODO: Technically we should separate the transition from all-row constraints
+                    // assert_eq!(c.domain(), ConstraintDomain::EveryFrame(2));
```
Good Idea

adr1anh replied:
Done #392

Al-Kindi-0 replied:
Thank you

Al-Kindi-0 commented on codegen/ace/src/layout.rs:
````diff
@@ -0,0 +1,230 @@
+    ///   ```
+    /// - Each trace must be padded with zero columns such that each row is word-aligned.
+    ///
+    /// # TODO(adr1anh):
+    /// The degree of the quotient is fixed to 8 matching the degree of the VM constraints, but
+    /// the actual degree can be derived from the [`Air`].
````
If we have opened an issue for this, then we can probably just remove this, unless you have a strong preference for keeping it

adr1anh replied:
#391

Al-Kindi-0 commented on codegen/ace/src/lib.rs:
```diff
@@ -1,21 +1,148 @@
+                match root.domain() {
+                    ConstraintDomain::FirstRow => boundary_first_roots.push(*root.node_index()),
+                    ConstraintDomain::LastRow => boundary_last_roots.push(*root.node_index()),
+                    _ => unreachable!("TODO"),
```
It seems not, so can we remove the TODO and replace with something more informative

Al-Kindi-0 reviewed (approved):
A few remaining minor points, and this should good to go.
Thank you again!

adr1anh commented:
Addressed the last remaining comments, thanks a lot for the thorough feedback!

adr1anh commented:
> Look great. However it's a pity that the changes to the regression files are all squashed, so they are useless. When moving files around it's better to leave them as isolated git commits so one can easily check that there were no modifications. When modifying the regressions (e.g. for syntax), those changes should be in a separate commits from semantics changes in the code.

You're right, there were clearly changes in the layout (especially due to the extra padding) which are worth comparing. When moving the files **and** regenerating the dot files, GitHub the similarity index was too low for GitHub to understand they were the same files (having similar names is not enough). For this reason, I've squashed all commits left the regression files in the same directory so they are comparable. 

The following command shows that the files increased on average 13.6%, which I'm assuming is due to the padding. 
```bash
git diff --numstat origin/next...HEAD -- '*.dot' | \
awk '{added += $1; deleted += $2; orig += ($1 - $1 + $2)} 
     END { total = orig + added - deleted; 
           printf "Average increase: %.1f%% across %d .dot files\n", 
           100 * (total - orig) / orig, NR }'
> Average increase: 13.6% across 13 .dot files
```

adr1anh commented:
The discrepancy was due to the order in which we multiplied the vanishing polynomials, which enables memoization of repeated operations. I also noticed that we now always include the constants zero and one, which are the only nodes that can be disconnected in the graph.

Al-Kindi-0 reviewed (approved):
Looks good!
